c2rust-analyze/src/rewrite/ty.rs

```diff
@@ -325,19 +326,38 @@ impl<'a, 'tcx> intravisit::Visitor<'tcx> for HirTyVisitor<'a, 'tcx> {
     fn nested_visit_map(&mut self) -> Self::Map {
         self.acx.tcx().hir()
     }
+
+    fn visit_stmt(&mut self, s: &'tcx hir::Stmt<'tcx>) {
+        match s.kind {
+            // A local with a user type annotation
+            StmtKind::Local(hir_local) if hir_local.ty.is_some() => {
+                let rw_lcx = LabeledTyCtxt::new(self.acx.tcx());
+                for (mir_local, mir_local_decl) in self.mir.local_decls.iter_enumerated() {
```

Comment:
Rather than make this an O(N^2) operation, we should build a map up front from `Span` to the MIR `Local` with that span.  This would also give us a chance to check for ambiguous cases where there are two locals with the same span (maybe due to macro tricks).

Reply:
does bd8f4a36 accomplish what you're asking for? I was sort of unsure what subset of Spans from HIR we wanted to build the map with and if there was some each way to iterate over all of those. in that commit I just limit it to the spans of `hir::Local`s

Reply:
That looks like it works, but seems a bit fragile since it relies on MIR locals being declared in a particular order.  What I was thinking was actually to just loop over the `mir.local_decls` and insert with `mir_local_decl.source_info.span` as the map key (not looking at HIR at all).  Then you can still do lookups exactly like you've got in that commit, since we expect the MIR local and the HIR local to have the same `Span`.

Reply:
I've made the simplification in a6596623